### PR TITLE
AVS-421_116 - Start structure and functionality for AssociatedTaxables

### PR DESCRIPTION
**Pull request by sjb9774**
## Summary
Associated Taxables are used by Gift Wrapping, WEEE Tax, and some 3rd
party extension to allow calculcating tax for orders on items that are
not technically "items" in the Magento sense. This commit begins the
basic functionality of persisting the Associated Taxable information
from the order step so it can be re-transmitted to Avalara at the time
of invoice processing.

## Description
1. The Associated Taxable persistence begins at in the [QuoteSubmitSuccessObserver](https://github.com/classyllama/ClassyLlama_AvaTax/pull/213/files#diff-4f27843eecde589e445672d4a8e3880d)
   1. Each item in the quote is iterated over and the associated taxables saved each in their own row within the `avatax_associated_taxables` table. An example of an Associated Taxable at the item level is gift wrapping.
   2. The quote is checked for Associated Taxables, if they exist then they are also saved to the `avatax_associated_taxables` table with the `associated_item_code` of 'quote' and a `NULL` `order_item_id` to distinguish them from item-level taxables. An example of a quote-level associated taxable is a printed gift card
2. We don't care about these Associated Taxables again until an invoice is created for the order, I've added code to the existing [Invoice `aroundSave` plugin](https://github.com/classyllama/ClassyLlama_AvaTax/pull/213/files#diff-6ba84efa2fc8cd533ca284b7bb1277fdR134) that gets the associated taxables tied to each item invoiced and all the quote level items if they've not yet been invoiced and fills their `invoice_id` column with the invoice's id to indicate they are now associated with that invoice and should be used when the AvaTax queue is processed.
3. I added code to the [getTaxRequestForSalesObject](https://github.com/classyllama/ClassyLlama_AvaTax/pull/213/files#diff-defa2a9dc80d5ddb4732d944799e9648R668) (which is used to build the request that will be sent to Avalara for a given invoice or credit memo when the queue is processed) that pulls in all Associated Taxables associated with that invoice or creditmemo and [builds them into lines](https://github.com/classyllama/ClassyLlama_AvaTax/pull/213/files#diff-791fb5aa1dfb929abcdbd42be1c6992eR248) that will be sent to Avalara.

#### Associated Taxables Example Data
avatax_associated_taxables 2019-01-29 10-08-10

## Things that need to be tested

- [ ] Different currency
- [ ] Multiple address checkout

### CODE REVIEW

**Comment by rsisco** [comment]
@sjb9774 - Is the "Note" in your PR comment still applicable?

**Comment by sjb9774** [comment]
@rsisco Yes, there are still a few specific scenarios I'd like to test that could impact a few places in the code but I think the bulk of the code review would not be affected.

File: Plugin/Sales/Model/Spi/CreditmemoResource.php
Comment (sjb9774) [inline Plugin/Sales/Model/Spi/CreditmemoResource.php]: Should be `$taxables`

File: Observer/QuoteSubmitSuccessObserver.php
Comment (sjb9774) [inline Observer/QuoteSubmitSuccessObserver.php]: What happens when a quote _is_ virtual and has quote-level associated taxables?

File: Plugin/Sales/Model/Spi/InvoiceResource.php
Comment (sjb9774) [inline Plugin/Sales/Model/Spi/InvoiceResource.php]: Should be `$taxables`

File: Helper/AssociatedTaxable.php
Comment (sjb9774) [inline Helper/AssociatedTaxable.php]: Address or remove this TODO

File: Model/Tax/AssociatedTaxableRepository.php
Comment (sjb9774) [inline Model/Tax/AssociatedTaxableRepository.php]: Add docblock

File: Observer/QuoteSubmitSuccessObserver.php
Comment (sjb9774) [inline Observer/QuoteSubmitSuccessObserver.php]: Add `$associatedTaxableHelper` to docblock

File: Model/Tax/Sales/Total/Quote/Tax.php
Comment (sjb9774) [inline Model/Tax/Sales/Total/Quote/Tax.php]: Why were any of these new arguments introduced to this file?

File: Api/SearchCriteria/JoinProcessor/Order.php
Comment (erikhansen) [inline Api/SearchCriteria/JoinProcessor/Order.php]: @sjb9774 Determine if this is still needed.

File: Helper/AssociatedTaxable.php
Comment (erikhansen) [inline Helper/AssociatedTaxable.php]: Change method name to something more specific, clarifying what this method does

File: Model/Tax/AssociatedTaxable.php
Comment (erikhansen) [inline Model/Tax/AssociatedTaxable.php]: Add DocBlock to this and all methods missing

File: Model/Tax/AssociatedTaxableRepository.php
Comment (erikhansen) [inline Model/Tax/AssociatedTaxableRepository.php]: Missing DocBlock

File: Model/Tax/AssociatedTaxableRepository.php
Comment (erikhansen) [inline Model/Tax/AssociatedTaxableRepository.php]: Combine this method with `getAllAssociatedTaxablesForCreditMemo` and then pass in the type

File: Observer/QuoteSubmitSuccessObserver.php
Comment (erikhansen) [inline Observer/QuoteSubmitSuccessObserver.php]: Wrap all of this in a try/catch, where the type you catch is `\Throwable`. Then log the error to the AvaTax logs with a high severity level (may be "critical"?).

File: Observer/QuoteSubmitSuccessObserver.php
Comment (erikhansen) [inline Observer/QuoteSubmitSuccessObserver.php]: Probably need to refactor this to use `\Magento\Quote\Model\Quote::getAllShippingAddresses` to support multiple addresses.

File: Setup/UpgradeSchema.php
Comment (erikhansen) [inline Setup/UpgradeSchema.php]: Move this to below the `order_id` column

File: Plugin/Sales/Model/Spi/CreditmemoResource.php
Comment (erikhansen) [inline Plugin/Sales/Model/Spi/CreditmemoResource.php]: Move all of this code into its own method. Wrap with a try/catch block (with error logging).

File: etc/di.xml
Comment (erikhansen) [inline etc/di.xml]: Look at removing this code, through line 84

**Review by erikhansen — CHANGES_REQUESTED** [review]
@sjb9774 Per our discussion, I've added a number of items you'll need to address.

**Comment by erikhansen** [comment]
For the record, future work on this PR was put on hold, awaiting approval from Avalara to move forward on it. 

Internal task link: AVS-421_116

**Comment by erikhansen** [comment]
Closing this PR, per [my comment](https://github.com/classyllama/ClassyLlama_AvaTax/issues/207#issuecomment-480974662) in the related issue.

**Pull request closed without merging**
